### DISCUSSION TITLE
How best to send keys with modifiers

### DISCUSSION THREAD

**Discussion by hraftery** [NicoHood/HID#434] [Q&A]
I'm getting great value out of the library thank you, allowing me to send arbitrary HID Keyboard and Consumer reports from an UNO derivative.

My use case is sending particular keycodes, rather than characters. Before using the library I was packing the report structure manually, and now I'm struggling to see how to set (and in particular, unset) the modifier field effectively.

So suppose I start with three bytes: the modifier field (`m`) and two keycodes (`c1` and `c2`), and I want to send a press and release event. What's the best way to do this? Here's what I've come up with:

```
//Essentially, undo the packing in DefaultKeyboardAPI.hpp:33 so it will repack.
for(uint8_t k = KEY_LEFT_CTRL; k <= KEY_RIGHT_GUI; k++)
{
  if(m & bit(k-KEY_LEFT_CTRL))
    Keyboard.add((KeyboardKeycode)k);
}

//Then add the keycodes and send them
Keyboard.add(KeyboardKeycode(c1));
//Keyboard.write(KeyboardKeycode(c2)); //note "write" means add, press and release.
//Alas, write doesn't clear the modifiers on release. So do it one by one instead:
Keyboard.press(KeyboardKeycode(c2));
for(uint8_t k = KEY_LEFT_CTRL; k <= KEY_RIGHT_GUI; k++)
  Keyboard.remove((KeyboardKeycode)k);
Keyboard.releaseAll();
```

This feels super clumsy and hard won. I feel like I could improve, either by using the library better, or getting access to the report fields directly some how. Any ideas?

**Comment by CoderPlusPlus** [comment]
If I understand correctly, I think you're just thinking a little too low level for the way this project is implemented.  The following is how you would send "CTRL+ALT+DEL":


```
//press each key but don't release until each has been pressed
Keyboard.press(KEY_LEFT_CONTROL);
Keyboard.press(KEY_LEFT_ALT);
Keyboard.press(KEY_DELETE);
Keyboard.send(); //this is the key.  Gotta send that all 3 keys were pressed and not released.

//now that all three are pressed, release them!
Keyboard.release(KEY_LEFT_CONTROL);
Keyboard.release(KEY_LEFT_ALT);
Keyboard.release(KEY_DELETE);
Keyboard.send();
```

I'm using this project to interface a vintage (1988!) laptop AT keyboard to a modern system for a project I'm working on.  I'm able to do key combinations without issue using code similar to this, so I expect it should work for you.  I wish the documentation on this project was better.  It's really neat.  But it leaves us having to try to interpret the source to figure out how to use it, which sucks.  Hope that helps!

**Reply by hraftery** [reply]
> I think you're just thinking a little too low level for the way this project is implemented.

I think you're right there! Problem is I can't figure out how to use the abstraction to make the low level do what I want :-D

In your example you know the keys you're pressing. I'm just being handed the keys and need to send them, whatever they are. So I loop through the keys I've been given and call `add()` on each one. For my purposes using `press()` is redundant ([because it adds **and** sends](https://github.com/NicoHood/HID/blob/b16be57caef4295c6cd382a7e4c64db5073647f7/src/HID-APIs/KeyboardAPI.hpp#L60)) since one event with all buttons is sufficient. For similar reasons I think your `send()` is redundant.

So the `add()`/`press()` thing aside, I think our examples are the same. Same for the release stuff. You send 4 events, one for each key release, while I send one with all keys released. Yours is a bit more "realistic" (how do you release 4 keys simultaneously?) but the device I'm controlling doesn't seem to care. Fundamentally though, the strategy is the same - iterate through the keys adding or removing each one individually.

I think you've probably hit the nail on the head though - if you think in terms of a physical key-down-or-up-one-at-a-time situation, the abstraction starts to make sense. Me, I'm thinking "send a report with these keys" and getting a bit tied up.

Still, unless I'm missing something, even in a physical-key scenario `releaseAll()` not releasing modifiers seems like a bug. And it makes the misleadingly named `write()` a bit useless. I think the reasons are steeped in legacy compatibility, but I guess I was just surprised to find this so hard to figure out and feel like I'm missing something.
